Changes for Microsoft.DataMigration GA stable version 2018-04-19.

binuj opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

binuj committed:
Initial commit for GA stable version 2018-04-19. Files copied from 'preview/2018-03-31-preview'. Next commit will have the actual changes for version 2018-04-19.

binuj committed:
Changes for GA stable version 2018-04-19. Removed MI sceanrio as it is not scoped for GA.

AutorestCI commented:
# Automation for azure-libraries-for-java

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-go

binuj committed:
Part of GA stable version 2018-04-19 : Fixed autorest validation errors.

hitenjava commented on specification/datamigration/resource-manager/readme.md:
````diff
@@ -92,6 +92,25 @@ input-file:
 - Microsoft.DataMigration/preview/2018-03-31-preview/definitions/MigrationValidation.json
 - Microsoft.DataMigration/preview/2018-03-31-preview/definitions/ValidateMigrationInputSqlServerSqlMITask.json
 ```
+
````
Line 31 in readme.md has default as "tag: package-2018-03-31-preview".
Change this to the GA version. The public repo sdk picks up the version number from here.

binuj replied:
Thanks, done.

binuj committed:
Changed default package to 2018-04-19 in readme.md.

hovsepm commented on specification/datamigration/resource-manager/readme.md:
````diff
@@ -92,6 +92,25 @@ input-file:
 - Microsoft.DataMigration/preview/2018-03-31-preview/definitions/ValidateMigrationInputSqlServerSqlMITask.json
 ```
+
+### Tag: package-2018-04-19
````
please reorganize readme.md to have most resent tag at the top.

hovsepm commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/definitions/Common.json:
```diff
@@ -0,0 +1,310 @@
+          ],
+          "x-ms-enum": {
+            "name": "NameCheckFailureReason",
+            "modelAsString": true
```
>"modelAsString": true [](start = 12, length = 21)

just a reminder - if enum is not modeled as string any time you will need to add a new enum member to it will be a breaking change and will require an API version bump. I'd strongly recommend marking enums as string

hitenjava replied:
Just confirming - There is no change needed right? The usage here follows the recommended way of defining enums, correct?

binuj replied:
Currently all the enums are marked as "modelAsString": true

hovsepm replied:
No changes asked for now, just a warning for a possible breaking change in the future.

hovsepm replied:
My comment rather came from already existing enums in Tasks.json (e.g. ProjectTaskProperties.State). Your PR has all the new enums marked as string types which is exactly what I was recommending :) sorry for the confusion.

hovsepm commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/definitions/Common.json:
```diff
@@ -0,0 +1,310 @@
+          "description": "The current value of the quota. If null or missing, the current value cannot be determined in the context of the request."
+        },
+        "id": {
+          "type": "string",
```
Isn't this field read-only? if so should be marked as `"readOnly": true,`

hovsepm replied:
In general any property of a type that is supposed to be sent only in one direction from Server to Client should be marked as `"readOnly": true,`

---
In reply to: [183830418](https://github.com/Azure/azure-rest-api-specs/pull/2936#discussion_r183830418) [](ancestors = 183830418)

hovsepm commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/definitions/TasksCommon.json:
```diff
@@ -0,0 +1,601 @@
+      "type": "object",
+      "description": "Information about a single database",
+      "properties": {
+        "id": {
```
>id [](start = 9, length = 2)

Readonly?

hovsepm commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/definitions/TasksCommon.json:
```diff
@@ -0,0 +1,601 @@
+          "type": "string",
+          "description": "Unique identifier for the database"
+        },
+        "name": {
```
>name [](start = 9, length = 4)

Readonly?

binuj committed:
Addressed comments.
Ordered readme.md contents from latest to oldest version.
Marked response properties as readonly.

hovsepm commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/datamigration.json:
```diff
@@ -0,0 +1,1239 @@
+          "GET"
+        ],
+        "operationId": "Projects_List",
+        "x-ms-examples": {
```
Operation ID should be "Projects_ListByResourceGroup"

binuj replied:
Done.

hovsepm commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/definitions/MigrateSqlServerSqlDbTask.json:
```diff
@@ -0,0 +1,392 @@
+      ]
+    },
+    "MigrateSqlServerSqlDbTaskOutputDatabaseLevelValidationResult": {
+      "x-ms-discriminator-value": "MigrationDatabaseLevelValidationOutput",
```
Please add "description" for this type

binuj replied:
Done.

hovsepm commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/definitions/MigrateSqlServerSqlDbTask.json:
```diff
@@ -0,0 +1,392 @@
+      ]
+    },
+    "MigrateSqlServerSqlDbTaskOutputError": {
+      "type": "object",
```
Please add "description" for this type

binuj replied:
Done.

hovsepm commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/definitions/MigrateSqlServerSqlDbTask.json:
```diff
@@ -0,0 +1,392 @@
+      ]
+    },
+    "MigrateSqlServerSqlDbTaskOutputTableLevel": {
+      "type": "object",
```
Please add "description" for this type

binuj replied:
Done.

hovsepm commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/definitions/MigrateSqlServerSqlDbTask.json:
```diff
@@ -0,0 +1,392 @@
+      ]
+    },
+    "MigrateSqlServerSqlDbTaskOutputDatabaseLevel": {
+      "type": "object",
```
Please add "description" for this type

binuj replied:
Done.

hovsepm commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/definitions/MigrateSqlServerSqlDbTask.json:
```diff
@@ -0,0 +1,392 @@
+      "discriminator": "resultType"
+    },
+    "MigrateSqlServerSqlDbTaskOutputMigrationLevel": {
+      "x-ms-discriminator-value": "MigrationLevelOutput",
```
Please add "description" for this type

binuj replied:
Done.

hovsepm commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/definitions/TasksCommon.json:
```diff
@@ -0,0 +1,601 @@
+    "StartMigrationScenarioServerRoleResult": {
+      "type": "object",
+      "description": "",
+      "properties": {
```
Please add "description" for this type

binuj replied:
Done.

ravbhatnagar commented:
The only big change from ARM standpoint is making some properties readonly. Since this change is made in a new api version, and the older version was preview, its fine. signing off!

binuj committed:
Addressed comments. Added description for some types.

binuj commented on specification/datamigration/resource-manager/Microsoft.DataMigration/stable/2018-04-19/definitions/MigrateSqlServerSqlDbTask.json:
```diff
@@ -0,0 +1,392 @@
+      ]
+    },
+    "MigrateSqlServerSqlDbTaskOutputValidationResult": {
+      "x-ms-discriminator-value": "MigrationValidationOutput",
```
Done.

binuj commented on specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2017-11-15-preview/definitions/Tasks.json:
```diff
@@ -24,7 +24,8 @@
-          "type": "string"
+          "type": "string",
+          "readOnly": true
         },
```
Removed readonly.

binuj committed:
Fixed Linter errors for files in version '2018-03-31-preview'.

binuj committed:
Fixed Linter error.

hovsepm commented:
one left
```
Executing: npx autorest@2.0.4152 --validation --azure-validator --message-format=json specification/datamigration/resource-manager/readme.md --tag=package-2017-11-15-preview
{
  "type": "Error",
  "code": "RequiredReadOnlyProperties",
  "message": "Property 'taskType' is a required property. It should not be marked as 'readonly'.",
  "id": "R2056",
  "validationCategory": "SDKViolation",
  "providerNamespace": null,
  "resourceType": null,
  "sources": [
    "file:///home/travis/build/Azure/azure-rest-api-specs/specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2017-11-15-preview/definitions/Tasks.json:25:8 ($.definitions.ProjectTaskProperties.properties.taskType)"
  ],
  "jsonref": "file:///home/travis/build/Azure/azure-rest-api-specs/specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2017-11-15-preview/definitions/Tasks.json:25:8 ($.definitions.ProjectTaskProperties.properties.taskType)",
  "json-path": "file:///home/travis/build/Azure/azure-rest-api-specs/specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2017-11-15-preview/definitions/Tasks.json:25:8 ($.definitions.ProjectTaskProperties.properties.taskType)"
}
```

hovsepm commented:
There are also bunch of missing documentation fields. They will be used in all the auto-generated languages as hints for developers so it is highly suggested to have human readable documentation text. You can see all the doc issues in this build https://travis-ci.org/Azure/azure-rest-api-specs/jobs/372255731. They start with
```
{
  "type": "Warning",
  "code": "DescriptionAndTitleMissing",
```

binuj committed:
Addressed comnments. Added missing documentation.

hovsepm commented:
@binuj please revert changes to the specs of the previous versions - no need to fix those issues in this PR.

binuj committed:
Revert all breaking changes in preview version folders as this PR is only focused on stable 2018-04-19.
Marked some properties as readonly in  '2018-04-19\definitions\TasksCommon.json'

hovsepm commented:
@binuj please fix Linter errors like `"Property named: 'Logins', for definition: 'ConnectToTargetSqlMITaskOutput' must follow camelCase style. Example: 'logins'.", `

hovsepm commented:
ignore my last comment - errors are in previous preview specs and were marked by Linter due to description change in your current pr.
